decentralized-message-queue/src/DMQ/Protocol/LocalMsgSubmission/Type.hs

```diff
@@ -0,0 +1,33 @@
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE PolyKinds         #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+-- | This module provides the type of LocalMsgProtocol via LocalTxSubmission
+--
+module DMQ.Protocol.LocalMsgSubmission.Type
+  ( module DMQ.Protocol.LocalMsgSubmission.Type
+    -- * re-exports
+  , module Core
+  , module Ouroboros
+  ) where
+
+import Data.Typeable
+import Network.TypedProtocol.Core as Core
+import Ouroboros.Network.Protocol.LocalTxSubmission.Type as Ouroboros
+import Ouroboros.Network.Util.ShowProxy
+
+-- | The LocalMsgSubmission protocol is an alias for the LocalTxSubmission
+--
+type LocalMsgSubmission sig reason = Ouroboros.LocalTxSubmission sig (SigMempoolFail reason)
+
+-- | The type of failures when adding to the mempool
+--
+data SigMempoolFail reason =
+    SigInvalid reason
+  | SigDuplicate
+  | SigExpired
+  | SigResultOther reason
+  deriving (Eq, Show)
```

Comment:
We don't really need polymorphism here, don't we? And we're anyway forced to use `reason ~ Text` (or `String`).